src/components/implementation/simple_vmm/vmm/vcpuid.c

```diff
@@ -0,0 +1,763 @@
+	uint32_t *eax, uint32_t *ebx, uint32_t *ecx, uint32_t *edx)
+{
+	const struct vcpuid_entry *entry_check = find_vcpuid_entry(0x80000000U, 0);
+	uint64_t guest_ia32_misc_enable = 0x800001;
+	uint32_t leaf = 0x80000001U;
+
+	if ((entry_check != NULL) && (entry_check->eax >= leaf)) {
+		cpuid_subleaf(leaf, 0x0U, eax, ebx, ecx, edx);
+		/* SDM Vol4 2.1, XD Bit Disable of MSR_IA32_MISC_ENABLE
+		 * When set to 1, the Execute Disable Bit feature (XD Bit) is disabled and the XD Bit
+		 * extended feature flag will be clear (CPUID.80000001H: EDX[20]=0)
+		 */
+		if ((guest_ia32_misc_enable & MSR_IA32_MISC_ENABLE_XD_DISABLE) != 0UL) {
+			*edx = *edx & ~CPUID_EDX_XD_BIT_AVIL;
+		}
+	} else {
+		*eax = 0U;
+		*ebx = 0U;
+		*ecx = 0U;
+		*edx = 0U;
+	}
+}
+
+static void guest_limit_cpuid(const struct vmrt_vm_vcpu *vcpu, uint32_t leaf,
+	uint32_t *eax, uint32_t *ebx, uint32_t *ecx, uint32_t *edx)
+{
+	uint64_t guest_ia32_misc_enable = 0x800001;
+
+	if ((guest_ia32_misc_enable & MSR_IA32_MISC_ENABLE_LIMIT_CPUID) != 0UL) {
+		/* limit the leaf number to 2 */
+		if (leaf == 0U) {
+			*eax = 2U;
+		} else if (leaf > 2U) {
+			*eax = 0U;
+			*ebx = 0U;
+			*ecx = 0U;
+			*edx = 0U;
+		} else {
+			/* In this case, leaf is 1U, return the cpuid value get above */
+		}
+	}
+}
+
+void guest_cpuid(struct vmrt_vm_vcpu *vcpu, uint32_t *eax, uint32_t *ebx, uint32_t *ecx, uint32_t *edx)
+{
+	uint32_t leaf = *eax;
+	uint32_t subleaf = *ecx;
+
+	/* vm related */
+	if (!is_percpu_related(leaf)) {
+		const struct vcpuid_entry *entry = find_vcpuid_entry(leaf, subleaf);
+
+		if (entry != NULL) {
+			*eax = entry->eax;
+			*ebx = entry->ebx;
+			*ecx = entry->ecx;
+			*edx = entry->edx;
+		} else {
+			*eax = 0U;
+			*ebx = 0U;
+			*ecx = 0U;
+			*edx = 0U;
+		}
+	} else {
+		/* percpu related */
+		switch (leaf) {
+		case 0x01U:
+			guest_cpuid_01h(vcpu, eax, ebx, ecx, edx);
+			break;
+
+		case 0x0bU:
+			guest_cpuid_0bh(eax, ebx, ecx, edx);
+			break;
+
+		case 0x0dU:
+			guest_cpuid_0dh(eax, ebx, ecx, edx);
+			break;
+
+		case 0x19U:
+			guest_cpuid_19h(eax, ebx, ecx, edx);
+			break;
+
+		case 0x80000001U:
+			guest_cpuid_80000001h(vcpu, eax, ebx, ecx, edx);
+			break;
+
+		default:
+			/*
+			 * In this switch statement, leaf 0x01/0x0b/0x0d/0x19/0x80000001
+			 * shall be handled specifically. All the other cases
+			 * just return physical value.
+			 */
+			cpuid_subleaf(leaf, *ecx, eax, ebx, ecx, edx);
+			break;
+		}
+	}
+
+	guest_limit_cpuid(vcpu, leaf, eax, ebx, ecx, edx);
+}
+
+void 
+cpuid_handler(struct vmrt_vm_vcpu *vcpu)
+{
+	uint64_t rax, rbx, rcx, rdx;
+	volatile struct vm_vcpu_shared_region *regs = vcpu->shared_region;
+	int test = 0;
+
+	rax = regs->ax;
+	rbx = regs->bx;
+	rcx = regs->cx;
+	rdx = regs->dx;
+
+	guest_cpuid(vcpu, (uint32_t *)&rax, (uint32_t *)&rbx, (uint32_t *)&rcx, (uint32_t *)&rdx);
+	
+	regs->ax = rax;
+	regs->bx = rbx;
+	regs->cx = rcx;
+	regs->dx = rdx;
+
+	GOTO_NEXT_INST(regs);
+
+	return;
+}
+
```

Comment:
The complexity of all of this is a little sad. But I'm glad you were able to find other code to do this for you!